From 17ad599c70471981b6ddf15e0726b9327cc0fcda Mon Sep 17 00:00:00 2001
From: T H <socialentry@gmail.com>
Date: Mon, 16 Jul 2018 00:25:42 -0400
Subject: [PATCH 1/2] klte-common: Add HFR/HSR support

* Add 60fps and 120fps to wrapper
* Clean up media_profiles.xml, add HSR profiles and increase audio
  bitrates.
* If torch mode is on, turn it back on after preview/parameter
  changes, currently the vendor shuts it off when preview is stopped
  and it doesn't turn back on after start preview. In Snap this allows
  the torch to stay on while the user changes settings.

Change-Id: Ia44d6c6a970b95db7b9687e9b2391c828e5af230
---
 camera/CameraWrapper.cpp   | 177 +++++++++-
 configs/media_profiles.xml | 664 +++++++++++++++++++------------------
 2 files changed, 511 insertions(+), 330 deletions(-)

diff --git a/camera/CameraWrapper.cpp b/camera/CameraWrapper.cpp
index ee67107..db63928 100644
--- a/camera/CameraWrapper.cpp
+++ b/camera/CameraWrapper.cpp
@@ -49,6 +49,9 @@ using namespace android;
 static Mutex gCameraWrapperLock;
 static camera_module_t *gVendorModule = 0;
 
+static int hfr = 0;
+static int hsr = 0;
+
 static camera_notify_callback gUserNotifyCb = NULL;
 static camera_data_callback gUserDataCb = NULL;
 static camera_data_timestamp_callback gUserDataCbTimestamp = NULL;
@@ -117,7 +120,10 @@ static int check_vendor_module()
     return rv;
 }
 
+#define KEY_VIDEO_HFR "video-hfr"
 #define KEY_VIDEO_HFR_VALUES "video-hfr-values"
+#define KEY_VIDEO_HSR "video-hsr"
+#define KEY_VIDEO_FAST_FPS_MODE "fast-fps-mode"
 
 // nv12-venus is needed for blobs, but
 // framework has no idea what it is
@@ -130,6 +136,96 @@ static bool is_4k_video(CameraParameters &params) {
     return video_width * video_height == 3840 * 2160;
 }
 
+static void setHfrParameters(struct camera_device * device, bool reset) {
+    int id = CAMERA_ID(device);
+    if ((!hfr && !hsr) || (id == FRONT_CAMERA_ID)) return;
+
+    VENDOR_CALL(device, cancel_auto_focus);
+    VENDOR_CALL(device, stop_preview);
+
+    CameraParameters params;
+    params.unflatten(String8(fixed_set_params[id]));
+    params.set(CameraParameters::KEY_RECORDING_HINT, "false");
+
+    const char *flashMode = params.get(CameraParameters::KEY_FLASH_MODE);
+    bool isTorch = flashMode && !strcmp(flashMode, CameraParameters::FLASH_MODE_TORCH);
+    if (isTorch)
+        params.set(CameraParameters::KEY_FLASH_MODE, CameraParameters::FLASH_MODE_OFF);
+
+    free(fixed_set_params[id]);
+    fixed_set_params[id] = strdup(params.flatten().string());
+    VENDOR_CALL(device, set_parameters, fixed_set_params[id]);
+
+    if (reset) {
+        params.set(KEY_PHASE_AF, ON);
+        params.set(KEY_VIDEO_FAST_FPS_MODE, "0");
+        params.set(CameraParameters::KEY_PREVIEW_FPS_RANGE, "30000,30000");
+    } else {
+        // must set phase-af to off to fix hfr/hsr focusing
+        params.set(KEY_PHASE_AF, OFF);
+        if (hfr) {
+            switch (hfr) {
+                case 1:
+                    params.set(KEY_VIDEO_HFR, "60");
+                    params.set(KEY_VIDEO_FAST_FPS_MODE, "1");
+                    params.set(CameraParameters::KEY_PREVIEW_FPS_RANGE, "60000,60000");
+                    params.set(CameraParameters::KEY_PICTURE_SIZE, "1920x1080");
+                    break;
+                case 2:
+                    params.set(KEY_VIDEO_HFR, "120");
+                    params.set(KEY_VIDEO_FAST_FPS_MODE, "2");
+                    params.set(CameraParameters::KEY_PREVIEW_FPS_RANGE, "120000,120000");
+                    params.set(CameraParameters::KEY_PICTURE_SIZE, "1280x720");
+                    break;
+            }
+        }
+        if (hsr) {
+            switch (hsr) {
+                case 1:
+                    params.set(KEY_VIDEO_HSR, "60");
+                    params.set(KEY_VIDEO_FAST_FPS_MODE, "0");
+                    params.set(CameraParameters::KEY_PREVIEW_FPS_RANGE, "60000,60000");
+                    params.set(CameraParameters::KEY_PICTURE_SIZE, "1920x1080");
+                    break;
+                case 2:
+                    params.set(KEY_VIDEO_HSR, "120");
+                    params.set(KEY_VIDEO_FAST_FPS_MODE, "1");
+                    params.set(CameraParameters::KEY_PREVIEW_FPS_RANGE, "120000,120000");
+                    params.set(CameraParameters::KEY_PICTURE_SIZE, "1280x720");
+                    break;
+            }
+        }
+    }
+
+    params.set(CameraParameters::KEY_RECORDING_HINT, "true");
+
+    if (isTorch)
+        params.set(CameraParameters::KEY_FLASH_MODE, CameraParameters::FLASH_MODE_TORCH);
+
+    free(fixed_set_params[id]);
+    fixed_set_params[id] = strdup(params.flatten().string());
+    VENDOR_CALL(device, set_parameters, fixed_set_params[id]);
+
+    VENDOR_CALL(device, start_preview);
+}
+
+static void setTorch(struct camera_device *device, bool on) {
+    int id = CAMERA_ID(device);
+    if ((id == BACK_CAMERA_ID) && fixed_set_params[id]) {
+        CameraParameters params;
+        params.unflatten(String8(fixed_set_params[id]));
+        const char *flashMode = params.get(CameraParameters::KEY_FLASH_MODE);
+        bool isTorch = flashMode && !strcmp(flashMode, CameraParameters::FLASH_MODE_TORCH);
+        if (isTorch) {
+            params.set(CameraParameters::KEY_FLASH_MODE, (on ?
+                CameraParameters::FLASH_MODE_TORCH : CameraParameters::FLASH_MODE_OFF));
+            free(fixed_set_params[id]);
+            fixed_set_params[id] = strdup(params.flatten().string());
+            VENDOR_CALL(device, set_parameters, fixed_set_params[id]);
+        }
+    }
+}
+
 static char *camera_fixup_getparams(int __attribute__((unused)) id,
     const char *settings)
 {
@@ -155,19 +251,36 @@ static char *camera_fixup_getparams(int __attribute__((unused)) id,
                 videoSizes);
     }
 
-    /* If the vendor has HFR values but doesn't also expose that
-     * this can be turned off, fixup the params to tell the Camera
-     * that it really is okay to turn it off.
-     */
-    const char *hfrModeValues = params.get(KEY_VIDEO_HFR_VALUES);
-    if (hfrModeValues && !strstr(hfrModeValues, "off")) {
-        char hfrModes[strlen(hfrModeValues) + 4 + 1];
-        sprintf(hfrModes, "%s,off", hfrModeValues);
-        params.set(KEY_VIDEO_HFR_VALUES, hfrModes);
+    // set hfr values
+    if (id == BACK_CAMERA_ID) {
+        params.set(KEY_VIDEO_HFR_VALUES, "60,120,off");
+        params.set(KEY_VIDEO_HFR, OFF);
+        params.set(KEY_VIDEO_HSR, OFF);
+        params.set(CameraParameters::KEY_SUPPORTED_PREVIEW_FRAME_RATES, "15,24,30,60,120");
+
+        switch (hfr) {
+            case 1:
+                params.set(KEY_VIDEO_HFR, "60");
+                break;
+            case 2:
+                params.set(KEY_VIDEO_HFR, "120");
+                break;
+        }
+
+        switch (hsr) {
+            case 1:
+                params.set(KEY_VIDEO_HSR, "60");
+                break;
+            case 2:
+                params.set(KEY_VIDEO_HSR, "120");
+                break;
+        }
     }
 
-    /* Enforce video-snapshot-supported to true */
-    if (videoMode) {
+    /* Enforce video-snapshot-supported to true
+       Ignore for HFR modes, taking a snapshot during HFR currently stops the video from saving
+        */
+    if (videoMode && (!hfr && !hsr)) {
         params.set(CameraParameters::KEY_VIDEO_SNAPSHOT_SUPPORTED, "true");
     }
 
@@ -225,7 +338,37 @@ static char *camera_fixup_setparams(int id, const char *settings)
             /* need to translate video-hdr to rt-hdr */
             const char *vhdr = params.get(KEY_QC_VIDEO_HDR);
             params.set(KEY_QC_RT_HDR, vhdr && !strcmp(vhdr, "on") ? ON : OFF);
+
+            // handle HFR setting
+            const char* videoHfr = params.get(KEY_VIDEO_HFR);
+            if (videoHfr && strcmp(videoHfr, OFF) != 0) {
+                if (strcmp(videoHfr,"120") == 0) {
+                    hfr = 2;
+                } else if (strcmp(videoHfr,"60") == 0) {
+                    hfr = 1;
+                }
+                hsr = 0;
+            } else {
+                hfr = 0;
+            }
+
+            // handle HSR setting
+            const char* videoHsr = params.get(KEY_VIDEO_HSR);
+            if (videoHsr && strcmp(videoHsr, OFF) != 0) {
+                if (strcmp(videoHsr,"120") == 0) {
+                    hsr = 2;
+                } else if (strcmp(videoHsr,"60") == 0) {
+                    hsr = 1;
+                }
+                hfr = 0;
+            } else {
+                hsr = 0;
+            }
         }
+    } else {
+        // not supported in front camera
+        params.set(KEY_PHASE_AF, OFF);
+        params.set(KEY_DYNAMIC_RANGE_CONTROL, OFF);
     }
 
     ALOGV("%s: Fixed parameters:", __FUNCTION__);
@@ -344,6 +487,8 @@ static int camera_start_preview(struct camera_device *device)
     ALOGV("%s->%08X->%08X", __FUNCTION__, (uintptr_t)device,
             (uintptr_t)(((wrapper_camera_device_t*)device)->vendor));
 
+    setTorch(device, true);
+
     return VENDOR_CALL(device, start_preview);
 }
 
@@ -355,6 +500,8 @@ static void camera_stop_preview(struct camera_device *device)
     ALOGV("%s->%08X->%08X", __FUNCTION__, (uintptr_t)device,
             (uintptr_t)(((wrapper_camera_device_t*)device)->vendor));
 
+    setTorch(device, false);
+
     VENDOR_CALL(device, stop_preview);
 }
 
@@ -398,6 +545,8 @@ static int camera_start_recording(struct camera_device *device)
         camera_set_parameters(device, strdup(parameters.flatten().string()));
     }
 
+    setHfrParameters(device, false);
+
     return VENDOR_CALL(device, start_recording);
 }
 
@@ -409,6 +558,8 @@ static void camera_stop_recording(struct camera_device *device)
     ALOGV("%s->%08X->%08X", __FUNCTION__, (uintptr_t)device,
             (uintptr_t)(((wrapper_camera_device_t*)device)->vendor));
 
+    setHfrParameters(device, true);
+
     VENDOR_CALL(device, stop_recording);
 }
 
@@ -568,6 +719,10 @@ static int camera_device_close(hw_device_t *device)
         goto done;
     }
 
+    // set back to default so other apps run normally
+    hsr = 0;
+    hfr = 0;
+
     for (int i = 0; i < camera_get_number_of_cameras(); i++) {
         if (fixed_set_params[i])
             free(fixed_set_params[i]);
diff --git a/configs/media_profiles.xml b/configs/media_profiles.xml
index 6487b08..7b2acab 100644
--- a/configs/media_profiles.xml
+++ b/configs/media_profiles.xml
@@ -78,68 +78,68 @@
      on an android-powered device.
 -->
 <MediaSettings>
-    <!-- Each camcorder profile defines a set of predefined configuration parameters -->
-    <CamcorderProfiles cameraId="0">
-
-        <EncoderProfile quality="qvga" fileFormat="3gp" duration="60">
-            <Video codec="m4v"
-                   bitRate="128000"
-                   width="320"
-                   height="240"
-                   frameRate="15" />
-            <Audio codec="amrnb"
-                   bitRate="12200"
-                   sampleRate="8000"
-                   channels="1" />
-        </EncoderProfile>
-
-        <EncoderProfile quality="cif" fileFormat="mp4" duration="30">
-            <Video codec="h264"
-                   bitRate="1200000"
-                   width="352"
-                   height="288"
-                   frameRate="30" />
-            <Audio codec="aac"
-                   bitRate="96000"
-                   sampleRate="48000"
-                   channels="1" />
-        </EncoderProfile>
-
-        <EncoderProfile quality="480p" fileFormat="mp4" duration="30">
-            <Video codec="h264"
-                   bitRate="6000000"
-                   width="720"
-                   height="480"
-                   frameRate="30" />
-            <Audio codec="aac"
-                   bitRate="96000"
-                   sampleRate="48000"
-                   channels="1" />
-        </EncoderProfile>
-
-        <EncoderProfile quality="720p" fileFormat="mp4" duration="30">
-            <Video codec="h264"
-                   bitRate="12000000"
-                   width="1280"
-                   height="720"
-                   frameRate="30" />
-            <Audio codec="aac"
-                   bitRate="96000"
-                   sampleRate="48000"
-                   channels="1" />
-        </EncoderProfile>
-
-        <EncoderProfile quality="1080p" fileFormat="mp4" duration="30">
-            <Video codec="h264"
-                   bitRate="17000000"
-                   width="1920"
-                   height="1080"
-                   frameRate="30" />
-            <Audio codec="aac"
-                   bitRate="96000"
-                   sampleRate="48000"
-                   channels="1" />
-        </EncoderProfile>
+  <!-- Each camcorder profile defines a set of predefined configuration parameters -->
+  <CamcorderProfiles cameraId="0">
+
+    <EncoderProfile quality="qvga" fileFormat="3gp" duration="60">
+      <Video codec="m4v"
+             bitRate="128000"
+             width="320"
+             height="240"
+             frameRate="15" />
+      <Audio codec="amrnb"
+             bitRate="12200"
+             sampleRate="8000"
+             channels="1" />
+    </EncoderProfile>
+
+    <EncoderProfile quality="cif" fileFormat="mp4" duration="30">
+      <Video codec="h264"
+             bitRate="1200000"
+             width="352"
+             height="288"
+             frameRate="30" />
+      <Audio codec="aac"
+             bitRate="128000"
+             sampleRate="48000"
+             channels="1" />
+    </EncoderProfile>
+
+    <EncoderProfile quality="480p" fileFormat="mp4" duration="30">
+      <Video codec="h264"
+             bitRate="6000000"
+             width="720"
+             height="480"
+             frameRate="30" />
+      <Audio codec="aac"
+             bitRate="256000"
+             sampleRate="48000"
+             channels="1" />
+    </EncoderProfile>
+
+    <EncoderProfile quality="720p" fileFormat="mp4" duration="30">
+      <Video codec="h264"
+             bitRate="12000000"
+             width="1280"
+             height="720"
+             frameRate="30" />
+      <Audio codec="aac"
+             bitRate="256000"
+             sampleRate="48000"
+             channels="1" />
+    </EncoderProfile>
+
+    <EncoderProfile quality="1080p" fileFormat="mp4" duration="30">
+      <Video codec="h264"
+             bitRate="17000000"
+             width="1920"
+             height="1080"
+             frameRate="30" />
+      <Audio codec="aac"
+             bitRate="256000"
+             sampleRate="48000"
+             channels="1" />
+    </EncoderProfile>
 
     <EncoderProfile quality="2160p" fileFormat="mp4" duration="30">
       <Video codec="h264"
@@ -151,73 +151,73 @@
       <Audio codec="aac"
              bitRate="320000"
              sampleRate="48000"
-             channels="2" />
+             channels="1" />
     </EncoderProfile>
 
-        <EncoderProfile quality="timelapseqcif" fileFormat="mp4" duration="30">
-            <Video codec="h264"
-                   bitRate="192000"
-                   width="176"
-                   height="144"
-                   frameRate="30" />
-            <!-- audio setting is ignored -->
-            <Audio codec="amrnb"
-                   bitRate="12200"
-                   sampleRate="8000"
-                   channels="1" />
-        </EncoderProfile>
-
-        <EncoderProfile quality="timelapsecif" fileFormat="mp4" duration="30">
-            <Video codec="h264"
-                   bitRate="1200000"
-                   width="352"
-                   height="288"
-                   frameRate="30" />
-            <!-- audio setting is ignored -->
-            <Audio codec="aac"
-                   bitRate="96000"
-                   sampleRate="48000"
-                   channels="1" />
-        </EncoderProfile>
-
-        <EncoderProfile quality="timelapse480p" fileFormat="mp4" duration="30">
-            <Video codec="h264"
-                   bitRate="6000000"
-                   width="720"
-                   height="480"
-                   frameRate="30" />
-            <!-- audio setting is ignored -->
-            <Audio codec="aac"
-                   bitRate="96000"
-                   sampleRate="48000"
-                   channels="1" />
-        </EncoderProfile>
-
-        <EncoderProfile quality="timelapse720p" fileFormat="mp4" duration="30">
-            <Video codec="h264"
-                   bitRate="12000000"
-                   width="1280"
-                   height="720"
-                   frameRate="30" />
-            <!-- audio setting is ignored -->
-            <Audio codec="aac"
-                   bitRate="96000"
-                   sampleRate="48000"
-                   channels="1" />
-        </EncoderProfile>
-
-        <EncoderProfile quality="timelapse1080p" fileFormat="mp4" duration="30">
-            <Video codec="h264"
-                   bitRate="17000000"
-                   width="1920"
-                   height="1080"
-                   frameRate="30" />
-            <!-- audio setting is ignored -->
-            <Audio codec="aac"
-                   bitRate="96000"
-                   sampleRate="48000"
-                   channels="1" />
-        </EncoderProfile>
+    <EncoderProfile quality="timelapseqcif" fileFormat="mp4" duration="30">
+      <Video codec="h264"
+             bitRate="192000"
+             width="176"
+             height="144"
+             frameRate="30" />
+      <!-- audio setting is ignored -->
+      <Audio codec="amrnb"
+             bitRate="12200"
+             sampleRate="8000"
+             channels="1" />
+    </EncoderProfile>
+
+    <EncoderProfile quality="timelapsecif" fileFormat="mp4" duration="30">
+      <Video codec="h264"
+             bitRate="1200000"
+             width="352"
+             height="288"
+             frameRate="30" />
+      <!-- audio setting is ignored -->
+      <Audio codec="aac"
+             bitRate="128000"
+             sampleRate="48000"
+             channels="1" />
+    </EncoderProfile>
+
+    <EncoderProfile quality="timelapse480p" fileFormat="mp4" duration="30">
+      <Video codec="h264"
+             bitRate="6000000"
+             width="720"
+             height="480"
+             frameRate="30" />
+      <!-- audio setting is ignored -->
+      <Audio codec="aac"
+             bitRate="256000"
+             sampleRate="48000"
+             channels="1" />
+    </EncoderProfile>
+
+    <EncoderProfile quality="timelapse720p" fileFormat="mp4" duration="30">
+      <Video codec="h264"
+             bitRate="12000000"
+             width="1280"
+             height="720"
+             frameRate="30" />
+      <!-- audio setting is ignored -->
+      <Audio codec="aac"
+             bitRate="256000"
+             sampleRate="48000"
+             channels="1" />
+    </EncoderProfile>
+
+    <EncoderProfile quality="timelapse1080p" fileFormat="mp4" duration="30">
+      <Video codec="h264"
+             bitRate="17000000"
+             width="1920"
+             height="1080"
+             frameRate="30" />
+      <!-- audio setting is ignored -->
+      <Audio codec="aac"
+             bitRate="256000"
+             sampleRate="48000"
+             channels="1" />
+    </EncoderProfile>
 
     <EncoderProfile quality="timelapse2160p" fileFormat="mp4" duration="30">
       <Video codec="h264"
@@ -230,208 +230,234 @@
       <Audio codec="aac"
              bitRate="320000"
              sampleRate="48000"
-             channels="2" />
+             channels="1" />
+    </EncoderProfile>
+
+    <!-- Highspeed Profiles -->
+    <EncoderProfile quality="highspeedlow" fileFormat="mp4" duration="30">
+      <Video codec="h264" bitRate="384000" width="176" height="144" frameRate="60" />
+      <Audio codec="aac" bitRate="128000" sampleRate="48000" channels="1" />
+    </EncoderProfile>
+
+    <EncoderProfile quality="highspeedhigh" fileFormat="mp4" duration="30">
+      <Video codec="h264" bitRate="17000000" width="1920" height="1080" frameRate="60" />
+      <Audio codec="aac" bitRate="256000" sampleRate="48000" channels="1" />
+    </EncoderProfile>
+
+    <EncoderProfile quality="highspeed480p" fileFormat="mp4" duration="30">
+      <Video codec="h264" bitRate="3449000" width="720" height="480" frameRate="60" />
+      <Audio codec="aac" bitRate="256000" sampleRate="48000" channels="1" />
+    </EncoderProfile>
+
+    <EncoderProfile quality="highspeed720p" fileFormat="mp4" duration="30">
+      <Video codec="h264" bitRate="12000000" width="1280" height="720" frameRate="60" />
+      <Audio codec="aac" bitRate="256000" sampleRate="48000" channels="1" />
     </EncoderProfile>
 
-        <ImageEncoding quality="95" />
-        <ImageEncoding quality="80" />
-        <ImageEncoding quality="70" />
-        <ImageDecoding memCap="20000000" />
-
-    </CamcorderProfiles>
-
-    <CamcorderProfiles cameraId="1">
-
-        <EncoderProfile quality="qvga" fileFormat="3gp" duration="60">
-            <Video codec="m4v"
-                   bitRate="128000"
-                   width="320"
-                   height="240"
-                   frameRate="15" />
-            <Audio codec="amrnb"
-                   bitRate="12200"
-                   sampleRate="8000"
-                   channels="1" />
-        </EncoderProfile>
-
-        <EncoderProfile quality="cif" fileFormat="mp4" duration="30">
-            <Video codec="h264"
-                   bitRate="1200000"
-                   width="352"
-                   height="288"
-                   frameRate="30" />
-            <Audio codec="aac"
-                   bitRate="96000"
-                   sampleRate="48000"
-                   channels="1" />
-        </EncoderProfile>
-
-        <EncoderProfile quality="480p" fileFormat="mp4" duration="30">
-            <Video codec="h264"
-                   bitRate="6000000"
-                   width="720"
-                   height="480"
-                   frameRate="30" />
-            <Audio codec="aac"
-                   bitRate="96000"
-                   sampleRate="48000"
-                   channels="1" />
-        </EncoderProfile>
-
-        <EncoderProfile quality="720p" fileFormat="mp4" duration="30">
-            <Video codec="h264"
-                   bitRate="12000000"
-                   width="1280"
-                   height="720"
-                   frameRate="30" />
-            <Audio codec="aac"
-                   bitRate="96000"
-                   sampleRate="48000"
-                   channels="1" />
-        </EncoderProfile>
-
-        <EncoderProfile quality="1080p" fileFormat="mp4" duration="30">
-            <Video codec="h264"
-                   bitRate="17000000"
-                   width="1920"
-                   height="1080"
-                   frameRate="30" />
-            <Audio codec="aac"
-                   bitRate="96000"
-                   sampleRate="48000"
-                   channels="1" />
-        </EncoderProfile>
-
-        <EncoderProfile quality="timelapseqcif" fileFormat="mp4" duration="30">
-            <Video codec="h264"
-                   bitRate="192000"
-                   width="176"
-                   height="144"
-                   frameRate="30" />
-            <!-- audio setting is ignored -->
-            <Audio codec="amrnb"
-                   bitRate="12200"
-                   sampleRate="8000"
-                   channels="1" />
-        </EncoderProfile>
-
-        <EncoderProfile quality="timelapsecif" fileFormat="mp4" duration="30">
-            <Video codec="h264"
-                   bitRate="1200000"
-                   width="352"
-                   height="288"
-                   frameRate="30" />
-            <!-- audio setting is ignored -->
-            <Audio codec="aac"
-                   bitRate="96000"
-                   sampleRate="48000"
-                   channels="1" />
-        </EncoderProfile>
-
-        <EncoderProfile quality="timelapse480p" fileFormat="mp4" duration="30">
-            <Video codec="h264"
-                   bitRate="6000000"
-                   width="720"
-                   height="480"
-                   frameRate="30" />
-            <!-- audio setting is ignored -->
-            <Audio codec="aac"
-                   bitRate="96000"
-                   sampleRate="48000"
-                   channels="1" />
-        </EncoderProfile>
-
-        <EncoderProfile quality="timelapse720p" fileFormat="mp4" duration="30">
-            <Video codec="h264"
-                   bitRate="12000000"
-                   width="1280"
-                   height="720"
-                   frameRate="30" />
-            <!-- audio setting is ignored -->
-            <Audio codec="aac"
-                   bitRate="96000"
-                   sampleRate="48000"
-                   channels="1" />
-        </EncoderProfile>
-
-        <EncoderProfile quality="timelapse1080p" fileFormat="mp4" duration="30">
-            <Video codec="h264"
-                   bitRate="17000000"
-                   width="1920"
-                   height="1080"
-                   frameRate="30" />
-            <!-- audio setting is ignored -->
-            <Audio codec="aac"
-                   bitRate="96000"
-                   sampleRate="48000"
-                   channels="1" />
-        </EncoderProfile>
-
-        <ImageEncoding quality="95" />
-        <ImageEncoding quality="80" />
-        <ImageEncoding quality="70" />
-        <ImageDecoding memCap="20000000" />
-
-    </CamcorderProfiles>
-
-    <EncoderOutputFileFormat name="3gp" />
-    <EncoderOutputFileFormat name="mp4" />
-
-    <!--
+    <EncoderProfile quality="highspeed1080p" fileFormat="mp4" duration="30">
+      <Video codec="h264" bitRate="17000000" width="1920" height="1080" frameRate="60" />
+      <Audio codec="aac" bitRate="256000" sampleRate="48000" channels="1" />
+    </EncoderProfile>
+
+    <ImageEncoding quality="95" />
+    <ImageEncoding quality="80" />
+    <ImageEncoding quality="70" />
+    <ImageDecoding memCap="20000000" />
+
+  </CamcorderProfiles>
+
+  <CamcorderProfiles cameraId="1">
+
+    <EncoderProfile quality="qvga" fileFormat="3gp" duration="60">
+      <Video codec="m4v"
+             bitRate="128000"
+             width="320"
+             height="240"
+             frameRate="15" />
+      <Audio codec="amrnb"
+             bitRate="12200"
+             sampleRate="8000"
+             channels="1" />
+    </EncoderProfile>
+
+    <EncoderProfile quality="cif" fileFormat="mp4" duration="30">
+      <Video codec="h264"
+             bitRate="1200000"
+             width="352"
+             height="288"
+             frameRate="30" />
+      <Audio codec="aac"
+             bitRate="128000"
+             sampleRate="48000"
+             channels="1" />
+    </EncoderProfile>
+
+    <EncoderProfile quality="480p" fileFormat="mp4" duration="30">
+      <Video codec="h264"
+             bitRate="6000000"
+             width="720"
+             height="480"
+             frameRate="30" />
+      <Audio codec="aac"
+             bitRate="128000"
+             sampleRate="48000"
+             channels="1" />
+    </EncoderProfile>
+
+    <EncoderProfile quality="720p" fileFormat="mp4" duration="30">
+      <Video codec="h264"
+             bitRate="12000000"
+             width="1280"
+             height="720"
+             frameRate="30" />
+      <Audio codec="aac"
+             bitRate="256000"
+             sampleRate="48000"
+             channels="1" />
+    </EncoderProfile>
+
+    <EncoderProfile quality="1080p" fileFormat="mp4" duration="30">
+      <Video codec="h264"
+             bitRate="17000000"
+             width="1920"
+             height="1080"
+             frameRate="30" />
+      <Audio codec="aac"
+             bitRate="256000"
+             sampleRate="48000"
+             channels="1" />
+    </EncoderProfile>
+
+    <EncoderProfile quality="timelapseqcif" fileFormat="mp4" duration="30">
+      <Video codec="h264"
+             bitRate="192000"
+             width="176"
+             height="144"
+             frameRate="30" />
+      <!-- audio setting is ignored -->
+      <Audio codec="amrnb"
+             bitRate="12200"
+             sampleRate="8000"
+             channels="1" />
+    </EncoderProfile>
+
+    <EncoderProfile quality="timelapsecif" fileFormat="mp4" duration="30">
+      <Video codec="h264"
+             bitRate="1200000"
+             width="352"
+             height="288"
+             frameRate="30" />
+      <!-- audio setting is ignored -->
+      <Audio codec="aac"
+             bitRate="128000"
+             sampleRate="48000"
+             channels="1" />
+    </EncoderProfile>
+
+    <EncoderProfile quality="timelapse480p" fileFormat="mp4" duration="30">
+      <Video codec="h264"
+             bitRate="6000000"
+             width="720"
+             height="480"
+             frameRate="30" />
+      <!-- audio setting is ignored -->
+      <Audio codec="aac"
+             bitRate="256000"
+             sampleRate="48000"
+             channels="1" />
+    </EncoderProfile>
+
+    <EncoderProfile quality="timelapse720p" fileFormat="mp4" duration="30">
+      <Video codec="h264"
+             bitRate="12000000"
+             width="1280"
+             height="720"
+             frameRate="30" />
+      <!-- audio setting is ignored -->
+      <Audio codec="aac"
+             bitRate="256000"
+             sampleRate="48000"
+             channels="1" />
+    </EncoderProfile>
+
+    <EncoderProfile quality="timelapse1080p" fileFormat="mp4" duration="30">
+      <Video codec="h264"
+             bitRate="17000000"
+             width="1920"
+             height="1080"
+             frameRate="30" />
+      <!-- audio setting is ignored -->
+      <Audio codec="aac"
+             bitRate="256000"
+             sampleRate="48000"
+             channels="1" />
+    </EncoderProfile>
+
+    <ImageEncoding quality="95" />
+    <ImageEncoding quality="80" />
+    <ImageEncoding quality="70" />
+    <ImageDecoding memCap="20000000" />
+
+  </CamcorderProfiles>
+
+  <EncoderOutputFileFormat name="3gp" />
+  <EncoderOutputFileFormat name="mp4" />
+
+  <!--
          If a codec is not enabled, it is invisible to the applications
          In other words, the applications won't be able to use the codec
          or query the capabilities of the codec at all if it is disabled
     -->
-    <VideoEncoderCap name="h264" enabled="true"
-        minBitRate="64000" maxBitRate="50000000"
-        minFrameWidth="176" maxFrameWidth="3840"
-        minFrameHeight="144" maxFrameHeight="2160"
-        minFrameRate="15" maxFrameRate="30" />
-
-    <VideoEncoderCap name="h263" enabled="true"
-        minBitRate="64000" maxBitRate="2000000"
-        minFrameWidth="176" maxFrameWidth="800"
-        minFrameHeight="144" maxFrameHeight="480"
-        minFrameRate="15" maxFrameRate="30" />
-
-    <VideoEncoderCap name="m4v" enabled="true"
-        minBitRate="64000" maxBitRate="40000000"
-        minFrameWidth="176" maxFrameWidth="1920"
-        minFrameHeight="144" maxFrameHeight="1080"
-        minFrameRate="15" maxFrameRate="30" />
-
-    <AudioEncoderCap name="aac" enabled="true"
-        minBitRate="758" maxBitRate="288000"
-        minSampleRate="8000" maxSampleRate="48000"
-        minChannels="1" maxChannels="1" />
-
-    <AudioEncoderCap name="heaac" enabled="true"
-        minBitRate="8000" maxBitRate="64000"
-        minSampleRate="16000" maxSampleRate="48000"
-        minChannels="1" maxChannels="1" />
-
-    <AudioEncoderCap name="aaceld" enabled="true"
-        minBitRate="16000" maxBitRate="192000"
-        minSampleRate="16000" maxSampleRate="48000"
-        minChannels="1" maxChannels="1" />
-
-    <AudioEncoderCap name="amrwb" enabled="true"
-        minBitRate="6600" maxBitRate="23050"
-        minSampleRate="16000" maxSampleRate="16000"
-        minChannels="1" maxChannels="1" />
-
-    <AudioEncoderCap name="amrnb" enabled="true"
-        minBitRate="5525" maxBitRate="12200"
-        minSampleRate="8000" maxSampleRate="8000"
-        minChannels="1" maxChannels="1" />
-
-    <!--
+  <VideoEncoderCap name="h264" enabled="true"
+                   minBitRate="64000" maxBitRate="50000000"
+                   minFrameWidth="176" maxFrameWidth="3840"
+                   minFrameHeight="144" maxFrameHeight="2160"
+                   minFrameRate="15" maxFrameRate="120" />
+
+  <VideoEncoderCap name="h263" enabled="true"
+                   minBitRate="64000" maxBitRate="2000000"
+                   minFrameWidth="176" maxFrameWidth="800"
+                   minFrameHeight="144" maxFrameHeight="480"
+                   minFrameRate="15" maxFrameRate="120" />
+
+  <VideoEncoderCap name="m4v" enabled="true"
+                   minBitRate="64000" maxBitRate="40000000"
+                   minFrameWidth="176" maxFrameWidth="1920"
+                   minFrameHeight="144" maxFrameHeight="1080"
+                   minFrameRate="15" maxFrameRate="30" />
+
+  <AudioEncoderCap name="aac" enabled="true"
+                   minBitRate="758" maxBitRate="288000"
+                   minSampleRate="8000" maxSampleRate="48000"
+                   minChannels="1" maxChannels="1" />
+
+  <AudioEncoderCap name="heaac" enabled="true"
+                   minBitRate="8000" maxBitRate="64000"
+                   minSampleRate="16000" maxSampleRate="48000"
+                   minChannels="1" maxChannels="1" />
+
+  <AudioEncoderCap name="aaceld" enabled="true"
+                   minBitRate="16000" maxBitRate="192000"
+                   minSampleRate="16000" maxSampleRate="48000"
+                   minChannels="1" maxChannels="1" />
+
+  <AudioEncoderCap name="amrwb" enabled="true"
+                   minBitRate="6600" maxBitRate="23050"
+                   minSampleRate="16000" maxSampleRate="16000"
+                   minChannels="1" maxChannels="1" />
+
+  <AudioEncoderCap name="amrnb" enabled="true"
+                   minBitRate="5525" maxBitRate="12200"
+                   minSampleRate="8000" maxSampleRate="8000"
+                   minChannels="1" maxChannels="1" />
+
+  <!--
         FIXME:
         We do not check decoder capabilities at present
         At present, we only check whether windows media is visible
         for TEST applications. For other applications, we do
         not perform any checks at all.
     -->
-    <VideoDecoderCap name="wmv" enabled="false"/>
-    <AudioDecoderCap name="wma" enabled="false"/>
+  <VideoDecoderCap name="wmv" enabled="false"/>
+  <AudioDecoderCap name="wma" enabled="false"/>
 </MediaSettings>
-- 
2.17.1

